packages/node-core/src/indexer/dictionary.service.ts

```diff
@@ -250,45 +305,65 @@ export class DictionaryService {
       return undefined;
     }
 
-    const {query, variables} = this.dictionaryQuery(startBlock, queryEndBlock, batchSize, conditions);
+    const blockHeightSet = new Set<number>();
+    const entityEndBlock: {[entity: string]: number} = {};
+
+    let hasNextPage = true;
+    let afterCursor: string | undefined;
+
     try {
-      const resp = await timeout(
-        this.client.query({
-          query: gql(query),
-          variables,
-        }),
-        this.nodeConfig.dictionaryTimeout
-      );
-      const blockHeightSet = new Set<number>();
-      const entityEndBlock: {[entity: string]: number} = {};
-      for (const entity of Object.keys(resp.data)) {
-        if (entity !== '_metadata' && resp.data[entity].nodes.length >= 0) {
-          for (const node of resp.data[entity].nodes) {
-            blockHeightSet.add(Number(node.blockHeight));
-            entityEndBlock[entity] = Number(node.blockHeight); //last added event blockHeight
+      let _metadata: MetaData;
+
+      do {
+        const {query, variables} = this.dictionaryQuery(startBlock, queryEndBlock, batchSize, conditions, afterCursor);
+
+        const resp = await timeout(
+          this.client.query({
+            query: gql(query),
+            variables: {...variables, afterCursor},
+          }),
+          this.nodeConfig.dictionaryTimeout
+        );
+
+        for (const entity of Object.keys(resp.data)) {
+          if (entity !== '_metadata') {
+            const data = this.supportsCursor ? resp.data[entity].edges : resp.data[entity].nodes;
```

Comment:
Rather than having various checks for `supportsCursor` it would be more readable to have a single check and keep the differences separate